python/ray/dataframe/utils.py

```diff
@@ -12,6 +12,32 @@
 _NAN_BLOCKS = {}
 
 
+class memoize:
+    """A basic memoizer that cache the input and output of the remote function
+
+    Notes:
+        - This will not work if the argument is mutable. Use with caution.
+    """
+
+    def __init__(self, f):
+        self.old_remote_func = f.remote
+        self.cache = {}
+        self.object_table = ray.global_state.object_table
+
+    def remote(self, *args):
+        """Return cached result if the arguments are cached
+        """
+        args = tuple(args)
+
+        if args in self.cache:
+            cached_result = self.cache[args]
+            return cached_result
```

Comment:
We don't need to worry about the result being evicted because ray will re-construct it for the user during `ray.get` Unless it's a driver task, like `ray.put`, which can be a bit buggy right now (issue #2156)